### PR TITLE
graph.py code cleaning

### PR DESCRIPTION
**Pull request by louwersj**
changes to graph.py to refactored code to become PEP8 compliant. Requires more work in a later stage, however a first step.

### CODE REVIEW

File: fmdtools/analyze/graph.py [hunk 1]
```diff
@@ -681,6 +785,8 @@ def draw_pyvis(self, filename="graph", width=1000, filt=True, physics=False, not
             Whether to display sliders. The default is True.
         physics : Bool, optional
             Whether to use physics during node placement. The default is False.
+        notebook : Bool, optional
+            todo : requires documentation update.
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:789]: passes notebook arg to pyviz Network (for displaying in jupyter)

File: fmdtools/analyze/graph.py [hunk 2]
```diff
@@ -821,6 +936,7 @@ def plot_high_degree_nodes(self, p=90, title='', node_kwargs={'node_color':'red'
             percentile of degrees to return, between 0 and 100
         title : str, optional
             Title for the plot. The default is 'High Degree Nodes'.
+        node_kwargs : todo : needs documentation update
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:939]: dict : kwargs to overwrite the default NodeStyle

File: fmdtools/analyze/graph.py [hunk 3]
```diff
@@ -1032,96 +1159,182 @@ def data_average(data):
...
 
-def data_error(data,average):
-    """Calculates error for each column in data"""
+def data_error(data, average):
+    """
+    Calculates error for each column in data
+
+    Parameters
+    ----------
+    data
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1168]: data : list
   List of lists from sff_model
average : list
   Average of data generated from sff_model over time

Returns
----------
lower_error : float
    Lower bound of error 
upper_error : float 
    Upper bound of error

File: fmdtools/analyze/graph.py [hunk 4]
```diff
@@ -1032,96 +1159,182 @@ def data_average(data):
...
-        self.t=0
-        self.fig, (self.bax, self.ax) = plt.subplots(2, gridspec_kw={'height_ratios': [1,10]})
-        self.kwargs=kwargs
-        self.g_obj=g_obj
+        """
+
+        Parameters
+        ----------
+        g_obj
+        kwargs
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1212]: g_obj : Graph 
    Graph object to plot interactively
kwargs : dict
    kwargs for Graph.draw

File: fmdtools/analyze/graph.py [hunk 5]
```diff
@@ -1032,96 +1159,182 @@ def data_average(data):
...
         self.ax.grid(True, which='both')
         self.ax.set_title('Drag nodes to change their positions')
         self.t+=1
         plt.pause(0.001)
+
     def print_pos(self):
-        print({k:list(v) for k,v in self.g_obj.pos.items()})
+        """
+        TODO : Needs documentation
+        Returns
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1333]: Prints the current node positions in the graph so they can be viewed (and copied) from the console.

File: fmdtools/analyze/graph.py [hunk 6]
```diff
@@ -1147,41 +1360,99 @@ def __init__(self, mdl, withstates=True, **kwargs):
...
+
     def nx_from_obj(self, mdl):
-        g=mdl.graph.copy()
-        labels = {fname:f.get_typename() for fname, f in mdl.fxns.items()}
-        labels.update({fname:f.get_typename() for fname, f in mdl.flows.items()})
+        """
+        TODO Needs documentation
+        Parameters
+        ----------
+        mdl
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1372]: Generates the networkx.graph object corresponding to the model.

mdl: Model
     Model to create the graph representation of.

Returns
--------
g : networkx.Graph
     networkx.Graph representation of model functions and flows (along with their attributes)

File: fmdtools/analyze/graph.py [hunk 7]
```diff
@@ -1147,41 +1360,99 @@ def __init__(self, mdl, withstates=True, **kwargs):
...
+        """
+        g = mdl.graph.copy()
+        labels = {fname: f.get_typename() for fname, f in mdl.fxns.items()}
+        labels.update({fname: f.get_typename() for fname, f in mdl.flows.items()})
         nx.set_node_attributes(g, labels, name='label')
         nx.set_edge_attributes(g, 'contains', name='label')
         return g
+
     def set_nx_states(self, mdl):
+        """
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1386]: Attaches state attributes to Graph corresponding to the states of the model

mdl: Model
    Model to represent.

File: fmdtools/analyze/graph.py [hunk 8]
```diff
@@ -1147,41 +1360,99 @@ def __init__(self, mdl, withstates=True, **kwargs):
...
+        -------
+
+        """
         self.set_flow_nodestates(mdl)
         self.set_fxn_nodestates(mdl)
+
     def set_fxn_nodestates(self, mdl):
+        """
+        TODO : requires documentation
+        Parameters
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1402]: Attaches state attributes to Graph corresponding to the states of the model that belong to functions

mdl: Model
    Model to represent.

File: fmdtools/analyze/graph.py [hunk 9]
```diff
@@ -1147,41 +1360,99 @@ def __init__(self, mdl, withstates=True, **kwargs):
...
         fxnfaults, fxnstates = {}, {}
         for fxnname, fxn in mdl.fxns.items():
             fxnstates[fxnname] = asdict(mdl.fxns[fxnname].s)
             fxnfaults[fxnname] = [*mdl.fxns[fxnname].m.faults]
         nx.set_node_attributes(self.g, fxnstates, 'states')
         nx.set_node_attributes(self.g, fxnfaults, 'faults')
+
     def set_flow_nodestates(self, mdl):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1419]: Attaches node state attributes to Graph corresponding to the states of the model that belong to flows.

mdl: Model
    Model to represent.

File: fmdtools/analyze/graph.py [hunk 10]
```diff
@@ -1147,41 +1360,99 @@ def __init__(self, mdl, withstates=True, **kwargs):
...
-        flows={}
-        multgraph = nx.projected_graph(mdl.graph, subedges ,multigraph=True)
+
+    def get_multi_edges(self, mdl, subedges):
+        """
+        Used by subclasses to attach functions/flows (subedges arg) to edges
+
+        Parameters
+        ----------
+        mdl
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1440]: mdl: Model
    Model to represent.
subedges : list
    nodes from the full graph which will become edges in the subgraph (e.g. individual flows)

Returns
----------
flows : dict
     Dictionary of edges with keys representing each sub-attribute of the edge (e.g., flows)

File: fmdtools/analyze/graph.py [hunk 11]
```diff
@@ -1336,6 +1625,8 @@ def nx_from_obj(self, mdl, withflows = True, **kwargs):
 
         Parameters
         ----------
+        mdl : TODO Needs documentation update.
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1628]: mdl: Model
    Model to represent

File: fmdtools/analyze/graph.py [hunk 12]
```diff
@@ -1559,48 +1878,124 @@ class ASGGraph(Graph):
     def __init__(self, asg, withstates=True):
         self.g = nx.compose(asg.flow_graph, asg.action_graph) 
         self.set_nx_labels(asg)
-        if withstates: self.set_nx_states(asg)
+        if withstates:
+            self.set_nx_states(asg)
+
     def set_nx_labels(self, asg):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1886]: Labels the underlying networkx graph structure with type attributes corresponding to the ASG.

asg : ASG
    Action Sequence Graph object to represent

File: fmdtools/analyze/graph.py [hunk 13]
```diff
@@ -1559,48 +1878,124 @@ class ASGGraph(Graph):
...
-            if e in asg.action_graph.edges():   self.g.edges[e]['label'] = 'condition'
-            elif e in asg.flow_graph.edges():   self.g.edges[e]['label'] = 'contains' 
+            if e in asg.action_graph.edges():
+                self.g.edges[e]['label'] = 'condition'
+            elif e in asg.flow_graph.edges():
+                self.g.edges[e]['label'] = 'contains'
+
     def set_nx_states(self, asg):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1908]: Attaches state and fault information to the underlying graph.

asg : ASG
     Underlying action sequence graph object to get states from

File: fmdtools/analyze/graph.py [hunk 14]
```diff
@@ -1559,48 +1878,124 @@ class ASGGraph(Graph):
...
             states[aname] = asdict(action.s)
             faults[aname] = [*action.m.faults]
         for fname, flow in asg.flows.items():
             states[fname] = asdict(flow.s)
         nx.set_node_attributes(self.g, states, 'states')
         nx.set_node_attributes(self.g, faults, 'faults')
+
     def set_edge_labels(self, title='label', title2='', subtext='name', **edge_label_styles):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1932]: same as Graph.set_edge_labels

File: fmdtools/analyze/graph.py [hunk 15]
```diff
@@ -1559,48 +1878,124 @@ class ASGGraph(Graph):
...
+
+        Returns
+        -------
+
+        """
         super().set_edge_labels(title=title, title2=title2, subtext=subtext, **edge_label_styles)
+
     def set_node_styles(self, active={}, **node_styles):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1948]: same as Graph.set_node_styles

File: fmdtools/analyze/graph.py [hunk 16]
```diff
@@ -1559,48 +1878,124 @@ class ASGGraph(Graph):
...
+
+        Returns
+        -------
+
+        """
         super().set_node_styles(active=active, **node_styles)
+
     def draw_graphviz(self, layout="twopi", overlap='voronoi', **kwargs):
+        """
+        TODO : requires documentation
```
Comment (hulsed) [inline fmdtools/analyze/graph.py:1963]: Same as Graph.draw_graphviz

**Review by hulsed — CHANGES_REQUESTED** [review]
I added comments where you listed documentation as "TODO," so add those to the documentation and I will merge in the PR.

File: fmdtools/analyze/graph.py [hunk 17]
```diff
@@ -3,45 +3,52 @@
...
-#Contributors: Daniel Hulse, Sequoia Andrade, Hannah Walsh
-#Created: November 2019
-#Refactored: April 2020
-#Added major interfaces: July 2021
-#Adopted object-oriented architecture and merged in networks.py: April 2023
+
+# File Name: analyze/graph.py
+# Contributors: Daniel Hulse, Sequoia Andrade, Hannah Walsh, Johan Louwers
+# Created: November 2019
+# Refactored: MAY 2023
```
[the remark this answers is not in the document]
Comment (louwersj) [reply fmdtools/analyze/graph.py:25]: moving it back to APR 2020 than :-)

File: fmdtools/analyze/graph.py [hunk 18]
```diff
@@ -99,21 +111,23 @@ def as_gv_kwargs(self):
                      'active':      dict(node_color='green'),
                      'degraded':    dict(node_color='orange'),
                      'faulty':      dict(edgecolors='red'),
-                     'high_degree_nodes': dict(node_color='red'),
+                     'high_degree_nodes': dict(node_color='red'),  # TODO this looks to be double. Needs checking
```
[the remark this answers is not in the document]
Comment (louwersj) [reply fmdtools/analyze/graph.py:114]: will remove it.

File: fmdtools/analyze/graph.py [hunk 19]
```diff
@@ -481,6 +570,10 @@ def draw(self, figsize=(12,10), title="", fig=False, ax=False, withlegend=True,
             Size for the figure (plt.figure arg). The default is (12,10).
         title : str, optional
             Title for the plot. The default is "".
+        fig : bool, optional
+            TODO : Need documentation update
+        ax : bool, optional
+            TODO : need documentation update
```
[the remark this answers is not in the document]
Comment (louwersj) [reply fmdtools/analyze/graph.py:576]: done

File: fmdtools/analyze/graph.py (hunk 1 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:789]: Done

File: fmdtools/analyze/graph.py (hunk 4 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1212]: Done

File: fmdtools/analyze/graph.py (hunk 5 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1333]: Done

File: fmdtools/analyze/graph.py (hunk 6 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1372]: Done

File: fmdtools/analyze/graph.py (hunk 7 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1386]: Done

File: fmdtools/analyze/graph.py (hunk 8 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1402]: Done

File: fmdtools/analyze/graph.py (hunk 9 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1419]: Done

File: fmdtools/analyze/graph.py (hunk 10 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1440]: Done

File: fmdtools/analyze/graph.py (hunk 11 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1628]: Done

File: fmdtools/analyze/graph.py (hunk 12 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1886]: Done

File: fmdtools/analyze/graph.py (hunk 15 above)
Comment (louwersj) [reply fmdtools/analyze/graph.py:1948]: Done

**Comment by louwersj** [comment]
@hulsed taken all your comments into account and are now part of commit #2 of this pull request.

**Pull request merged**